cni-plugin/tests/calico_cni_k8s_test.go

```diff
@@ -2999,6 +3001,130 @@ var _ = Describe("Kubernetes CNI tests", func() {
+						Name: name,
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{{
+							Name:  name,
+							Image: "ignore",
+						}},
+						NodeName: hostname,
+					},
+				})
+
+			// Create the container, which will call CNI and by default it will create the container with interface name 'eth0'.
+			containerID, _, _, _, _, contNs, err := testutils.CreateContainer(netconf, name, testutils.K8S_TEST_NS, "")
+			Expect(err).Should(HaveOccurred())
+			// Make sure the pod gets cleaned up, whether we fail or not.
+			expectedIfaceName := "eth0"
+			defer func() {
+				_, err := testutils.DeleteContainerWithIdAndIfaceName(netconf, contNs.Path(), name, testutils.K8S_TEST_NS, containerID, expectedIfaceName)
+				Expect(err).ShouldNot(HaveOccurred())
+
+				ensurePodDeleted(clientset, testutils.K8S_TEST_NS, name)
+			}()
+
+		})
+	})
+
+	Context("using valid readiness_gates", func() {
+		// Create a test http endpoint
+		srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			_, err := w.Write([]byte("Ok"))
+			Expect(err).NotTo(HaveOccurred())
+		}))
+		testEndpoint := srv.URL
+		netconf := fmt.Sprintf(`
+				{
+				  "cniVersion": "%s",
+				  "name": "net10",
+				  "type": "calico",
+				  "etcd_endpoints": "http://%s:2379",
+				  "datastore_type": "%s",
+           			  "nodename_file_optional": true,
+				  "log_level": "info",
+				  "readiness_gates": ["%s"],
+			 	  "ipam": {
+				    "type": "calico-ipam"
+				  },
+				  "kubernetes": {
+				    "kubeconfig": "/home/user/certs/kubeconfig"
+				  },
+				  "policy": {"type": "k8s"}
+				}`, cniVersion, os.Getenv("ETCD_IP"), os.Getenv("DATASTORE_TYPE"), testEndpoint)
+
+		It("should successfully create container", func() {
+			// Create a new ipPool.
+			testutils.MustCreateNewIPPool(calicoClient, "10.0.0.0/24", false, false, true)
+
+			clientset := getKubernetesClient()
+
+			// Now create a K8s pod.
+			name := "mypod-1"
+
+			ensureNamespace(clientset, testutils.K8S_TEST_NS)
+			ensurePodCreated(clientset, testutils.K8S_TEST_NS,
+				&v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: name,
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{{
+							Name:  name,
+							Image: "ignore",
+						}},
+						NodeName: hostname,
+					},
+				})
+
+			// Create the container, which will call CNI and by default it will create the container with interface name 'eth0'.
+			containerID, _, _, _, _, contNs, err := testutils.CreateContainer(netconf, name, testutils.K8S_TEST_NS, "")
+			Expect(err).ShouldNot(HaveOccurred())
+			// Make sure the pod gets cleaned up, whether we fail or not.
+			expectedIfaceName := "eth0"
+			defer func() {
+				_, err := testutils.DeleteContainerWithIdAndIfaceName(netconf, contNs.Path(), name, testutils.K8S_TEST_NS, containerID, expectedIfaceName)
+				Expect(err).ShouldNot(HaveOccurred())
+				ensurePodDeleted(clientset, testutils.K8S_TEST_NS, name)
+				srv.Close()
```

Comment:
I think we need to move this `srv.Close()` into its own defer right after the server is created.

There's a chance that we don't ever reach this defer, which means we'll leak a server instance.

Reply:
Initially that is where I had put the close, but then the server was closed even before the subject node(It()) is called.
I am putting the close now inside AfterEach() and starting server in BeforeEach() , this could have been done in BeforeAll and AfterAll, but that seems to be supported only from ginkgo v2. Since we have only one subject node inside the context I think it would not make a difference. I will push the code, please let me know in case you think otherwise.

Reply:
Gotcha - the beforeeach / aftereach looks like a good approach for this then :+1:

Reply:
Gotcha - the beforeeach / aftereach looks like a good approach for this then :+1: